### PR TITLE
Add rapid test with bad proposal coming from byzantine node

### PR DESCRIPTION
**Pull request by begmaroman**
# Description

As part of rapid testing, the following scenario should be covered: (1/3-1) byzantine nodes send a bad message during a round -> these messages should be rejected and consensus should be reached. 

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have added sufficient documentation in code

### CODE REVIEW

File: core/rapid_test.go [hunk 1]
```diff
@@ -383,15 +395,209 @@ func TestProperty_MajorityHonestNodes(t *testing.T) {
...
+		// Make sure that the inserted proposal is valid for each height
+		for _, proposalMap := range insertedProposals.proposals {
+			for _, insertedProposal := range proposalMap {
+				assert.Equal(t, message.proposal, insertedProposal)
+			}
+		}
+	})
+}
+
+// TestProperty_MajorityHonestNodes_BadProposal is a property-based test
```
Comment (JekaMas) [inline core/rapid_test.go:415]: The malicious validator is always malicious, don't it? Could we do rather arbitrary? Particular rounds and stages?

Comment (begmaroman) [reply core/rapid_test.go:415]: Yes, I will, but I would make it work first. Currently, the consensus can't be reached even with only one malicious node which is weird. Trying to figure out why it happens but still stuck here.

File: core/rapid_test.go [hunk 2]
```diff
@@ -383,15 +395,209 @@ func TestProperty_MajorityHonestNodes(t *testing.T) {
...
+			}
+		}
+
+		// commonBackendCallback is the common method modification required
+		// for the Backend, for all nodes
+		commonBackendCallback := func(backend *mockBackend, nodeIndex int) {
+			message := correctMessage
+			if uint64(nodeIndex) < numByzantineNodes {
+				message = badMessage
+			}
```
Comment (begmaroman) [inline core/rapid_test.go:476]: (1/3 - 1) bad nodes send bad message during round

File: core/rapid_test.go [hunk 3]
```diff
@@ -383,15 +395,209 @@ func TestProperty_MajorityHonestNodes(t *testing.T) {
...
+			logger.errorFn = func(s string, i ...interface{}) {
+				fmt.Println(append([]interface{}{s}, i...)...)
+				t.Errorf(s, i...)
+			}
+
+			logger.debugFn = func(s string, i ...interface{}) {
+				fmt.Println(append([]interface{}{s}, i...)...)
+				t.Logf(s, i...)
+			}
+		}
```
Comment (begmaroman) [inline core/rapid_test.go:460]: For debugging, will remove it

File: core/rapid_test.go [hunk 4]
```diff
@@ -383,15 +395,209 @@ func TestProperty_MajorityHonestNodes(t *testing.T) {
...
+
+		// Run the sequence up until a certain height
+		for height := uint64(0); height < desiredHeight; height++ {
+			// Start the main run loops
+			cluster.runSequence(height)
+
+			// Wait until Quorum nodes finish their run loop
+			ctx, cancelFn := context.WithTimeout(context.Background(), time.Second*5)
+			err := cluster.awaitNCompletions(ctx, int64(quorum(numNodes)))
```
Comment (begmaroman) [inline core/rapid_test.go:598]: It just does not receive completions even with one bad node

File: core/rapid_test.go (hunk 2 above)
Comment (JekaMas) [reply core/rapid_test.go:460]: But single round. This is possible to write without property based testing.

File: core/rapid_test.go [hunk 5]
```diff
@@ -385,11 +515,53 @@ func TestProperty_MajorityHonestNodes(t *testing.T) {
...
+			} else {
+				for _, insertedProposal := range proposalMap {
+					assert.Equal(t, correctRoundMessage.proposal, insertedProposal)
+				}
 			}
 		}
 	})
 }
+
+func createCluster(
```
[the remark this answers is not in the document]
Comment (begmaroman) [reply core/rapid_test.go]: Good point, done

**Review by vcastellm — APPROVED** [review]
LGTM

File: core/rapid_test.go (hunk 2 above)
Comment (JekaMas) [reply core/rapid_test.go]: do we have a task to extend property based and have a random sequence of events normal and malicious?

**Review by b00ris — APPROVED** [review]

File: core/rapid_test.go (hunk 2 above)
Comment (begmaroman) [reply core/rapid_test.go]: Just created one: EVM-195

**Pull request merged**